pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,160 @@
 package cli
 
 import (
-	"bytes"
-	"errors"
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
 
 	"github.com/go-logr/logr"
 
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/kubernetes/scheme"
-	restclient "k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/remotecommand"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/medik8s/fence-agents-remediation/api/v1alpha1"
 )
 
 type Executer interface {
-	Execute(pod *corev1.Pod, command []string) (stdout string, stderr string, err error)
+	AsyncExecute(uid types.UID, command []string)
+	Exists(uid types.UID) bool
 }
 
 type executer struct {
-	log       logr.Logger
-	config    *restclient.Config
-	clientSet *kubernetes.Clientset
+	client.Client
+	ctx      context.Context
+	log      logr.Logger
+	routines map[types.UID]bool
+	mapLock  sync.Mutex
 }
 
-var _ Executer = executer{}
+var _ Executer = &executer{}
 
 // NewExecuter builds the executer
-func NewExecuter(config *restclient.Config) (Executer, error) {
+func NewExecuter(ctx context.Context, client client.Client) (Executer, error) {
 	logger := ctrl.Log.WithName("executer")
 
-	clientSet, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		logger.Error(err, "failed building k8s client")
-		return nil, err
-	}
-
 	return &executer{
-		log:       logger,
-		config:    config,
-		clientSet: clientSet,
+		Client:   client,
+		log:      logger,
+		ctx:      ctx,
+		routines: make(map[types.UID]bool),
```

Comment:
Is `mapLock` should not be initialized here as well?

Reply:
IIUC it is automatically initialized